src/pubsub.c

```diff
@@ -481,7 +591,111 @@ NULL
     } else if (!strcasecmp(c->argv[1]->ptr,"numpat") && c->argc == 2) {
         /* PUBSUB NUMPAT */
         addReplyLongLong(c,dictSize(server.pubsub_patterns));
+    } else if (!strcasecmp(c->argv[1]->ptr,"local") && c->argc >= 2) {
+        /* PUBSUB LOCAL CHANNELS [PATTERN] */
+        if (!strcasecmp(c->argv[2]->ptr,"channels")) {
+            sds pat = (c->argc == 3) ? NULL : c->argv[3]->ptr;
+            channel_list(c, pat, server.pubsublocal_channels);
+
+        } else if (!strcasecmp(c->argv[2]->ptr,"numsub")) {
+            /* PUBSUB LOCAL NUMSUB [Channel_1 ... Channel_N] */
+            int j;
+
+            addReplyArrayLen(c, (c->argc - 3) * 2);
+            for (j = 3; j < c->argc; j++) {
+                list *l = dictFetchValue(server.pubsublocal_channels, c->argv[j]);
+
+                addReplyBulk(c, c->argv[j]);
+                addReplyLongLong(c, l ? listLength(l) : 0);
+            }
+        } else {
+            addReplySubcommandSyntaxError(c);
+        }
     } else {
         addReplySubcommandSyntaxError(c);
     }
 }
+
+void channel_list(client *c, sds pat, dict *pubsub_channels) {
+    dictIterator *di = dictGetIterator(pubsub_channels);
+    dictEntry *de;
+    long mblen = 0;
+    void *replylen;
+
+    replylen = addReplyDeferredLen(c);
+    while((de = dictNext(di)) != NULL) {
+        robj *cobj = dictGetKey(de);
+        sds channel = cobj->ptr;
+
+        if (!pat || stringmatchlen(pat, sdslen(pat),
+                                   channel, sdslen(channel),0))
+        {
+            addReplyBulk(c,cobj);
+            mblen++;
+        }
+    }
+    dictReleaseIterator(di);
+    setDeferredArrayLen(c,replylen,mblen);
+}
+
+/* PUBLISHLOCAL <channel> <message> */
+void publishLocalCommand(client *c) {
+    if (pubsubCheckACLPermissionsOrReply(c,1,1,0) != ACL_OK) return;
+    pubsubmeta *meta = createPubSubLocalMeta(c);
+    int receivers = pubsubPublishMessageInternal(
+            c->argv[1],
+            c->argv[2],
+            server.pubsublocal_channels,
+            1);
+    forceCommandPropagation(c,PROPAGATE_REPL);
+    addReplyLongLong(c,receivers);
+    zfree(meta);
+}
+
+/* SUBSCRIBELOCAL channel [channel ...] */
+void subscribeLocalCommand(client *c) {
```

Comment:
I don't think you should be able to subscribe to multiple different slots, and I think that needs to be within the function.

Reply:
If it is within the same node, what is the issue we foresee? 

With the current implementation, the logic is a client can subscribe to multiple slots within the same node.

Reply:
It was already getting handled with `CROSSSLOT Keys` error, I've put an additional check and message `-CROSSSLOT Channels in request don't hash to the same slot`